B2C Manual Setup Readme (#191)
Azure/azure-saas: docs/azure-saas-docs/content/en/components/Identity/identity-framework-setup-manual.md

---
type: docs
title: "Identity Framework Setup - Manual"
weight: 5000
toc_hide: true
---

On this page, you will find instructions for how to manually setup the Identity Framework through a combination of the Azure dashboard as well as relevant PowerShell commands

# Setup Identity Framework - Manual (Advanced)

## 1. Requirements

- [Azure CLI](https://docs.microsoft.com/en-us/cli/azure/)
- [Open SSL](https://www.openssl.org/)
  - Note: An OpenSSL executable is also bundled with Git for Windows at `C:\Program Files\Git\usr\bin\openssl.exe`
- Visual Studio Code

>    - Tools:

Author: ```suggestion
   - Required Extensions:
```

   - [Azure AD B2C Custom Policy Extension](https://marketplace.visualstudio.com/items?itemName=AzureADB2CTools.aadb2c)

## 2. Instructions

### Step 1. Create B2C Tenant in Azure

![(Azure Add Service Screenshot)](/azure-saas/images/identity-framework-manual-step-1-adding-service.png)

- From the dashboard, click `Create a Resource` and select `Azure Active Directory B2C`
- Note: A reference is registered in your subscription at time of creation to associate the tenant with your subscription.

### Step 2. Log into newly created Tenant

![(Switch Directory Screenshot)](/azure-saas/images/identity-framework-manual-step-2-switch-tenant-quicklink.png)

> - Upon being created, your Azure AD B2C Tenant will create a federated identity with your Azure subscription account.

Author: I'm not sure how I feel about this sentence. It might be hurting more than helping.

This is technically known as Azure AD B2B, which opens up a whole other can of worms of things to explain. Probably easier to just remove it and not explain what's happening under the hood here.

- Go to the `Azure AD B2C` dashboard while logged into your new Tenant, click `Users` in the lefthand navbar
- After navigating, you may create or invite guest users into your Tenant allowing them access

### Step 4. Create necessary app registrations in Azure AD B2C Tenant

![(Registered Apps Screenshot)](/azure-saas/images/identity-framework-manual-step-4-app-registration.png)

- From the Tenant `Azure AD B2C` dashboard, click `App registrations`
- Add new registrations corresponding to the following list of modules

#### Modules

- asdk-admin-api
  - Display Name: asdk-admin-api
  - Account Type: Accounts in any identity provider or organizational directory
  - Grant Admin Consent to openid and offline_access_permissions: true
- asdk-signupadmin-app
  - Display Name: asdk-saas-app
  - Account Type: Accounts in any identity provider or organizational directory
  - Redirect URI: [Single-page application] https://{your-signupadmin-app-url}.com/signin-oidc
  - Grant Admin Consent to openid and offline_access_permissions: true
- asdk-saas-app
  - Display Name: asdk-saas-app
  - Account Type: Accounts in any identity provider or organizational directory

>   - Redirect URI: [Single-page application] https://{your-app-url}.com/signin-oidc

Author: `https://appapplication{providerName}{environmentName}.azurewebsites.net/signin-oidc`

  - Grant Admin Consent to openid and offline_access_permissions: true
- asdk-permissions-api
  - Display Name: asdk-permissions-api
  - Account Type: Accounts in any identity provider or organizational directory
  - Grant Admin Consent to openid and offline_access_permissions: true
- IdentityExperienceFramework
  - Display Name: IdentityExperienceFramework
  - Account Type: Accounts in any identity provider or organizational directory
  - Grant Admin Consent to openid and offline_access_permissions: true
- ProxyIdentityExperienceFramework
  - Display Name: ProxyIdentityExperienceFramework
  - Account Type: Accounts in any identity provider or organizational directory
  - Grant Admin Consent to openid and offline_access_permissions: true

#### Module Client Secrets

![(Client Secrets Screenshot)](/azure-saas/images/badgemeup-screenshot.png)

- From the `App registration` view
- For the apps listed below, navigate to them, then navigate to `Certificates & secrets` in their navbars
- Add a new client secret for later reference, giving it an appropriate expiration and description
- Modules Requiring Client Secrets:

>   - asdk-saas-app

Author: Let's move these requirements into the list above. This section should just be the instructions.

  - asdk-signupadmin-app

#### Module API Scopes

![(Expose an API Scopes)](/azure-saas/images/badgemeup-screenshot.png)

- From the `App registration` view
- For the apps listed below, navigate to them, then navigate to `Expose an API` in their navbars
- The first time you add a scope, you will be prompted to define the `Application ID URI`, do so and save the domain and identifier each independently for later reference
- Add the following scopes to their respective APIs:

- asdk-admin-api:

> 

Author: Same as above. Let's move all the requirements into the list above.

  | Scope Name | Description |
  | - | - |
  | tenant.read | Read a customer's own Tenant data |
  | tenant.global.read | Admin-level read permissions for all Tenants |
  | tenant.write | Alter a customer's own Tenant data |
  | tenant.global.write | Admin-level write permissions for all Tenants |
  | tenant.delete | Delete a customer's own Tenant data |
  | tenant.global.delete | Admin-level delete permissions for all Tenants |

#### Module API Permissions

> - From the `App registration` view

Author: Same as above. Let's move all the requirements into the list above.

- asdk-saas-app
  - tenant.read
- asdk-signupadmin-app
  - tenant.read
  - tenant.global.read
  - tenant.write
  - tenant.global.write
  - tenant.delete
  - tenant.global.delete
  - Delete tenant global

### Step 5. Create self-signed certificate

>Note: A self-signed certificate is not recommended for production environments. Read more [here](https://azure.github.io/azure-saas/components/identity/permissions-service/#authentication).

- Create an RSA self-signed certificate to be referenced and uploaded to your Azure AD B2C Tenant
- The .pfx file will be uploaded in a later step to your Azure AD B2C to allow it access to integrate with your Permissions API
- OpenSSL with [Powershell Core (v7.0+)](https://github.com/PowerShell/PowerShell) Example:

```
$CertificatePassword = Read-Host -AsSecureString

openssl req -newkey rsa:4096 -x509 -sha256 -days 365 -nodes -out certificate.crt -keyout certificate.key -subj         "/CN=*.azurewebsites.net"

$pswd = ConvertFrom-SecureString -SecureString $CertificatePassword

openssl pkcs12 -export -out selfSignedCertificate.pfx -inkey certificate.key -in certificate.crt -password pass:$pswd

> ```

Author: (line 305 below is where they'll need it)

> ```

Reviewer: Whoops, those two lines were casualties when you mentioned to axe the thumbprint generation. Restored.

![(Identity Experience Framework Screenshot)](/azure-saas/images/identity-framework-manual-step-6-identity-experience-framework.png)

- Find `Policy keys` in `Azure AD B2C Service` by navigating to `Identity Experience Framework`
- Click Add and select the following options to generate the system policy keys

![(Policy Keys Screenshot)](/azure-saas/images/identity-framework-manual-step-6-policy-keys.png)

- TokenSigningKeyContainer
  - Options: Generate
  - Name: TokenSigningKeyContainer
  - Key type: RSA
  - Key usage: Signature
- TokenEncryptionKeyContainer
  - Options: Generate
  - Name: TokenEncryptionKeyContainer
  - Key type: RSA
  - Key usage: Encryption
- RestApiClientCertificate
  - Options: Upload
  - Name: *.azurewebsites.net
  - File upload: (The .pfx created in step 5)
  - Password: (The password created in step 5)

### Step 7. Clone ASDK repo

- Clone the latest release of the [ASDK](https://github.com/Azure/azure-saas)

### Step 8. Initiate identity Bicep deployment with parameters from previous few steps

- Generate a standard [Bicep parameters file](https://docs.microsoft.com/en-us/azure/azure-resource-manager/bicep/parameter-files) 

> - Navigate to `Saas.identity/Saas.Identity.Iac` for the `az deploy` to work correctly

Author: ```suggestion
- Navigate to `Saas.Identity/Saas.Identity.IaC` for the `az deploy` to work correctly
```

> - Deploy identity Bicep with the following parameters

Author: Add the command here that they can copy/paste. You can find it on line 552 of the b2c-create script.

| PolicySettings:ProxyIdentityExperienceFrameworkAppId | (App ID of ProxyIdentityExperienceFramework app created in step 4) |
| PolicySettings:PermissionsAPIUrl | (URL for the permissions endpoint in the Permission module (e.g. (PermissionsApiFQDN)/api/CustomClaims/permissions)) | 
| PolicySettings:RolesAPIUrl | (URL for the roles endpoint in the Permission module (e.g. (PermissionsApiFQDN)/api/CustomClaims/roles)) |

- Open execute all text field (shortcut: ctl + shift + P)
- Execute `B2C Build All Policies`
- Navigate to `Environments` dropdown in VS Code Explorer for generated policies

### Step 10. Upload IEF Policies to Azure AD B2C Tenant

![(Policy Upload Screenshot)](/azure-saas/images/identity-framework-manual-step-1-adding-service.png)

- In your Azure dashboard, navigate to `Azure AD B2C` service, then to `Identity Framework Experience` in the nav bar
- Select `Upload custom policy` and upload each of the generated policy files

### Step 11. Create parameters.json for input into main bicep deployment

![(Parameters File Directory Screenshot)](/azure-saas/images/badgemeup-screenshot.png)

> - Generate a standard [Bicep parameters file](https://docs.microsoft.com/en-us/azure/azure-resource-manager/bicep/parameter-files) 

Author: ```suggestion
- Generate a standard [Bicep parameters file](https://docs.microsoft.com/en-us/azure/azure-resource-manager/bicep/parameter-files) using the parameters below
```

- Place file into `src\Saas.Identity\Saas.IdentityProvider\policies`

> - Add the following values to the `parameters` collection

Author: ```suggestion
```

### Step 2. Log into newly created Tenant

![(Switch Directory Screenshot)](/azure-saas/images/identity-framework-manual-step-2-switch-tenant-quicklink.png)

- Click on `Open B2C Tenant` quicklink to immediately change directory to your new Tenant

### Step 3. (Optional) Invite collaborators to the new Tenant

Note: Collaborators are other developers you wish to help manage your services, not customers

![(Tenant User Management Screenshot)](/azure-saas/images/identity-framework-manual-step-3-invite-collaborators.png)

- Go to the `Azure AD B2C` dashboard while logged into your new Tenant, click `Users` in the lefthand navbar
- After navigating, you may create or invite guest users into your Tenant allowing them access

### Step 4. Create necessary app registrations in Azure AD B2C Tenant

![(Registered Apps Screenshot)](/azure-saas/images/identity-framework-manual-step-4-app-registration.png)

- From the Tenant `Azure AD B2C` dashboard, click `App registrations`
- Add new registrations corresponding to the following list of modules
- Instructions for adding client secrets, API scopes, and API permissions are detailed below

> 

Author: Add a note here letting them know that they need to capture/copy down all IDs and secrets generated as part of this process for use later on in the setup.

#### Modules

- asdk-admin-api
  - Display Name: asdk-admin-api
  - Account Type: Accounts in any identity provider or organizational directory
  - Grant Admin Consent to openid and offline_access_permissions: true
  - API Scopes:

    | Scope Name | Description |
    | - | - |
    | tenant.read | Read a customer's own Tenant data |
    | tenant.global.read | Admin-level read permissions for all Tenants |
    | tenant.write | Alter a customer's own Tenant data |
    | tenant.global.write | Admin-level write permissions for all Tenants |
    | tenant.delete | Delete a customer's own Tenant data |
    | tenant.global.delete | Admin-level delete permissions for all Tenants |

- asdk-signupadmin-app
  - Display Name: asdk-saas-app
  - Account Type: Accounts in any identity provider or organizational directory
  - Redirect URI: [Single-page application] `https://appsignup{providerName}{environmentName}.azurewebsites.net/signin-oidc`
  - Grant Admin Consent to openid and offline_access_permissions: true
  - Create Client Secrets?: Yes

>   - Required API Permissions: 

Author: I would do something like this:

- Required API Permissions:
   - asdk-admin-api
       - tenant.read
       - tenant.global.read
       - tenant.write
       - tenant.global.write
       - tenant.delete
       - tenant.global.delete

    - tenant.read
- asdk-permissions-api
  - Display Name: asdk-permissions-api
  - Account Type: Accounts in any identity provider or organizational directory
  - Grant Admin Consent to openid and offline_access_permissions: true
- IdentityExperienceFramework
  - Display Name: IdentityExperienceFramework
  - Account Type: Accounts in any identity provider or organizational directory
  - Grant Admin Consent to openid and offline_access_permissions: true
- ProxyIdentityExperienceFramework
  - Display Name: ProxyIdentityExperienceFramework
  - Account Type: Accounts in any identity provider or organizational directory
  - Grant Admin Consent to openid and offline_access_permissions: true

#### Module Client Secrets

![(Client Secrets Screenshot)](/azure-saas/images/badgemeup-screenshot.png)

- From the `App registration` view
- For the apps listed below, navigate to them, then navigate to `Certificates & secrets` in their navbars
- Add a new client secret for later reference, giving it an appropriate expiration and description

#### Module API Scopes

![(Expose an API Scopes)](/azure-saas/images/badgemeup-screenshot.png)

- From the `App registration` view

> - For the apps listed below, navigate to them, then navigate to `Expose an API` in their navbars

Author: ```suggestion
- For the apps that need a scope exposed, listed above, navigate to them, then navigate to `Expose an API` in their navbars
```

  - tenant.delete
  - tenant.global.delete
  - Delete tenant global

### Step 5. Create self-signed certificate

>Note: A self-signed certificate is not recommended for production environments. Read more [here](https://azure.github.io/azure-saas/components/identity/permissions-service/#authentication).

- Create an RSA self-signed certificate to be referenced and uploaded to your Azure AD B2C Tenant
- The .pfx file will be uploaded in a later step to your Azure AD B2C to allow it access to integrate with your Permissions API
- OpenSSL with [Powershell Core (v7.0+)](https://github.com/PowerShell/PowerShell) Example:

```
$pswd= Read-Host -Prompt "Please enter a password to encrypt the self signed certificate with"

# Create a new certificate in .crt format with the subject name as *.azurewebsites.net. You could also provide a specific subject name of your permissions api.
openssl req -newkey rsa:4096 -x509 -sha256 -days 365 -nodes -out certificate.crt -keyout certificate.key -subj         "/CN=*.azurewebsites.net"

# Export the .crt and .key to a .pfx format
openssl pkcs12 -export -out selfSignedCertificate.pfx -inkey certificate.key -in certificate.crt -password pass:$pswd

> # Encode pfx to base 64 string to reference in step 13

Author: ```suggestion
# Encode pfx to base 64 string to reference in step 13. You may output this when you're ready to copy it with `Write-Host $pfxString`
```

- TokenEncryptionKeyContainer
  - Options: Generate
  - Name: TokenEncryptionKeyContainer
  - Key type: RSA
  - Key usage: Encryption
- RestApiClientCertificate
  - Options: Upload
  - Name: *.azurewebsites.net
  - File upload: (The .pfx created in step 5)
  - Password: (The password created in step 5)

### Step 7. Clone ASDK repo

- Clone the latest release of the [ASDK](https://github.com/Azure/azure-saas)

### Step 8. Initiate identity Bicep deployment with parameters from previous few steps

- Generate a standard [Bicep parameters file](https://docs.microsoft.com/en-us/azure/azure-resource-manager/bicep/parameter-files) using the parameters below
- Navigate to `Saas.Identity/Saas.Identity.IaC` for the `az deploy` to work correctly
- Deploy identity Bicep with the following parameters
  - You may reference or fill out `main.parameters.json` in the Saas.Identity.IaC directory
  - (Note: some parameters are referenced in step 13 and must match what is selected here, such as the chosen saasProviderName)
-      az deployment group create --name "IdentityBicepDeployment" --resource-group <#YourAzureResourceGroupName#> template-file ./main.bicep --parameters ./main.parameters.json

> #### Bicep Templates to Deploy

Author: You can get rid of this section. This is outlined elsewhere in the docs.

---
type: docs
title: "Identity Framework Setup - Manual"
weight: 5000
toc_hide: true
---

> On this page, you will find instructions for how to manually setup the Identity Framework through a combination of the Azure dashboard as well as relevant PowerShell commands

Reviewer 2: ```suggestion
On this page, you will find instructions for how to manually setup the Identity Framework through a combination of the Azure Portal as well as relevant PowerShell commands
```

# Setup Identity Framework - Manual (Advanced)

## 1. Requirements

- [Azure CLI](https://docs.microsoft.com/en-us/cli/azure/)
- [Open SSL](https://www.openssl.org/)
  - Note: An OpenSSL executable is also bundled with Git for Windows at `C:\Program Files\Git\usr\bin\openssl.exe`
- [Visual Studio Code](https://code.visualstudio.com/)
   - Required Extensions:
   - [Azure AD B2C Custom Policy Extension](https://marketplace.visualstudio.com/items?itemName=AzureADB2CTools.aadb2c)

## 2. Instructions

### Step 1. Create B2C Tenant in Azure

- From the dashboard, click `Create a Resource` and select `Azure Active Directory B2C`
- Note: A reference is registered in your subscription at time of creation to associate the tenant with your subscription.

![(Azure Add Service Screenshot)](/azure-saas/images/identity-framework-manual-step-1-adding-service.png)

---

> ### Step 2. Log into newly created Tenant

Reviewer 2: ```suggestion
### Step 2. Log into newly created Azure B2C Tenant
```

> - Click on `Open B2C Tenant` quicklink to immediately change directory to your new Tenant

Reviewer 2: ```suggestion
- Click on `Open B2C Tenant` quicklink to immediately change directory to your new Azure B2C Tenant
```

![(Switch Directory Screenshot)](/azure-saas/images/identity-framework-manual-step-2-switch-tenant-quicklink.png)

---

> ### Step 3. (Optional) Invite collaborators to the new Tenant

Reviewer 2: ```suggestion
### Step 3. (Optional) Invite collaborators to the new Azure AD B2C tenant
```

Note: Collaborators are other developers you wish to help manage your services, not customers

> - Go to the `Azure AD B2C` dashboard while logged into your new Tenant, click `Users` in the lefthand navbar

Reviewer 2: ```suggestion
- Go to the `Azure AD B2C` dashboard while logged into your new Azure AD B2C tenant, click `Users` in the lefthand navbar
```

> - After navigating, you may create or invite guest users into your Tenant allowing them access

Reviewer 2: ```suggestion
- After navigating, you may create or invite guest users into your Azure AD B2C tenant allowing them access
```

![(Tenant User Management Screenshot)](/azure-saas/images/identity-framework-manual-step-3-invite-collaborators.png)

---

> ### Step 4. Create necessary app registrations in Azure AD B2C Tenant

Reviewer 2: ```suggestion
### Step 4. Create necessary app registrations in Azure AD B2C tenant
```

- From the Tenant `Azure AD B2C` dashboard, click `App registrations`

> - Add new registrations corresponding to the following list of modules

Reviewer 2: ```suggestion
- Add new registrations corresponding for the following list of modules
```

- Instructions for adding client secrets, API scopes, and API permissions are detailed below

> - Copy down all IDs and secret values for reference later on in the setup steps

Reviewer 2: ```suggestion
- Copy all client ids and secret values for reference later on in the setup steps
```

![(Registered Apps Screenshot)](/azure-saas/images/identity-framework-manual-step-4-app-registration.png)

#### Modules

- **asdk-admin-api**
  - Display Name: asdk-admin-api
  - Account Type: Accounts in any identity provider or organizational directory
  - Grant Admin Consent to openid and offline_access_permissions: true
  - API Scopes:

| Scope Name | Description |
| - | - |
| tenant.read | Read a customer's own Tenant data |
| tenant.global.read | Admin-level read permissions for all Tenants |
| tenant.write | Alter a customer's own Tenant data |
| tenant.global.write | Admin-level write permissions for all Tenants |
| tenant.delete | Delete a customer's own Tenant data |
| tenant.global.delete | Admin-level delete permissions for all Tenants |

- **asdk-signupadmin-app**

>   - Display Name: asdk-saas-app

Reviewer 2: ```suggestion
  - Display Name: asdk-signupadmin-app
```

Author: Good catch, thank you!

  - Redirect URI: [Single-page application] `https://appapplication{providerName}{environmentName}.azurewebsites.net/signin-oidc`
  - Grant Admin Consent to openid and offline_access_permissions: true
  - Create Client Secrets?: Yes
  - Required API Permissions: 
    - asdk-admin-api
      - tenant.read
- **asdk-permissions-api**
  - Display Name: asdk-permissions-api
  - Account Type: Accounts in any identity provider or organizational directory
  - Grant Admin Consent to openid and offline_access_permissions: true
- **IdentityExperienceFramework**
  - Display Name: IdentityExperienceFramework
  - Account Type: Accounts in any identity provider or organizational directory
  - Grant Admin Consent to openid and offline_access_permissions: true
- **ProxyIdentityExperienceFramework**
  - Display Name: ProxyIdentityExperienceFramework
  - Account Type: Accounts in any identity provider or organizational directory
  - Grant Admin Consent to openid and offline_access_permissions: true

#### Module Client Secrets

- From the `App registration` view
- For the apps that need a secret created (listed above), navigate to them, then navigate to `Certificates & secrets` in their navbars

> - Add a new client secret for later reference, giving it an appropriate expiration and description

Reviewer 2: ```suggestion
- Add a new client secret and copy it for later reference, giving it an appropriate expiration and description
```

| PolicySettings:IdentityExperienceFrameworkAppId | (App ID of IdentityExperienceFramework app created in step 4) |
| PolicySettings:ProxyIdentityExperienceFrameworkAppId | (App ID of ProxyIdentityExperienceFramework app created in step 4) |
| PolicySettings:PermissionsAPIUrl | (URL for the permissions endpoint in the Permission module (e.g. (PermissionsApiFQDN)/api/CustomClaims/permissions)) | 
| PolicySettings:RolesAPIUrl | (URL for the roles endpoint in the Permission module (e.g. (PermissionsApiFQDN)/api/CustomClaims/roles)) |

- Open the VS Code command palette (shortcut: ctrl + shift + p)
- Execute `B2C Build All Policies`
- Navigate to `Environments` dropdown in VS Code Explorer for generated policies

![(VS Code Screenshot)](/azure-saas/images/identity-framework-manual-step-9-vscode-ief-policies.png)

---

### Step 10. Upload IEF Policies to Azure AD B2C Tenant

- In your Azure dashboard, navigate to `Azure AD B2C` service, then to `Identity Framework Experience` in the nav bar
- Select `Upload custom policy` and upload each of the generated policy files

![(Policy Upload Screenshot)](/azure-saas/images/identity-framework-manual-step-10-upload-policies.png)

---

> ### Step 11. Create parameters.json for input into main bicep deployment

Reviewer 2: ```suggestion
### Step 11. Create parameters.json for input into main Bicep deployment
```

- Generate a standard [Bicep parameters file](https://docs.microsoft.com/en-us/azure/azure-resource-manager/bicep/parameter-files) using the parameters below

> - Place file into any folder (you will need it on step 2.b of the [Quick Start](https://azure.github.io/azure-saas/quick-start/#2b-deploying-to-azure---entire-solution) guide

Reviewer 2: ```suggestion
- Place file into any folder (you will need it on step 2.b of the [Quick Start](https://azure.github.io/azure-saas/quick-start/#2b-deploying-to-azure---entire-solution) guide)
```

#### Parameters

| Parameter Name | Value |
| - | - |
| adminApiScopes| (space delimited string of admin scope names (e.g. "test.scope tenant.delete tenant.global.delete tenant.global.read tenant.global.write tenant.read tenant.write"))|
| adminApiScopeBaseUrl| (Reference your Tenant domain and the registered Admin app identifier, format of| "https|//{your tenant name}.onmicrosoft.com/{admin app client id}")|
| azureAdB2cAdminApiClientIdSecretValue| (Admin Api Client Id found under its Registered App entry) |
| azureAdB2cDomainSecretValue| (Your Tenant domain found on your AD B2C dashboard)|

> | azureAdB2cInstanceSecretValue| (The B2C login endpoint in format of| https|//(Tenant Name).b2clogin.com)|

Reviewer 2: ```suggestion
| azureAdB2cInstanceSecretValue| (The B2C login endpoint in format of https://{TENANT_NAME}.b2clogin.com)|
```
